updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,140 @@
+        self.conn.write_all(b"downloading");
+    }
+
+    pub fn rebooting(&mut self) {
+        self.conn.write_all(b"rebooting");
+    }
+
+    pub fn cancel(&mut self) {
+        self.conn.write_all(b"cancel");
+    }
+
+    pub fn proceed() {}
+}
+
+impl StateChangeListener {
+    pub fn new() -> Self {
+        StateChangeListener { listeners: HashMap::new(), error_handlers: Vec::new() }
+    }
+
+    pub fn on(&mut self, state: StateID, f: fn(State)) {
+        let name = match state {
+            StateID::Download => String::from("download"),
+            StateID::Install => String::from("install"),
+            StateID::Reboot => String::from("reboot"),
+            StateID::Error => String::from("error"),
+        };
+        self.listeners.insert(name, f);
+    }
+
+    fn on_error(&mut self, f: fn(String)) {
+        self.error_handlers.push(f);
+    }
+
+    fn emit(&self, conn: UnixStream, state: &str) {
+        self.listeners.iter().for_each(|(_, f)| {
+            let conn = conn.clone();
+            f(State {
+                id: match state {
+                    "download" => StateID::Download,
+                    "install" => StateID::Install,
+                    "reboot" => StateID::Reboot,
+                    "error" => StateID::Error,
+                    _ => panic!("error"),
+                },
+                conn,
+            })
+        });
+    }
+
+    fn throw_error(&self, error: &str) {
+        self.error_handlers.iter().for_each(|f| {
+            let error = &error;
+            f(error.to_string())
+        })
+    }
+
+    async fn listen(&self) -> Result<()> {
+        let file = std::path::Path::new(&SDK_TRIGGER_FILENAME);
+        let listener = if file.exists() {
+            UnixListener::bind(format!("/run/{}", SOCKET_PATH)).await?
+        } else {
+            println!(
+                "WARNING: updatehub-sdk-statechange-trigger not found on {}",
+                SDK_TRIGGER_FILENAME
+            );
+            UnixListener::bind(SOCKET_PATH).await?
+        };
+
+        let ctrlc = CtrlC::new()?;
+        ctrlc
+            .race(async {
+                loop {
+                    process::exit(1);
+                }
+            })
+            .await;
+
+        loop {
+            let (socket, ..) = listener.accept().await?;
+            self.handle_connection(socket).await?;
+        }
+    }
+
+    async fn handle_connection(&self, stream: UnixStream) -> Result<()> {
+        let mut reader = BufReader::new(stream.clone());
+        let mut line = String::new();
+
+        loop {
+            let s = stream.clone();
+            let bytes = reader.read_line(&mut line).await?;
+
+            if bytes.to_string().contains("error") {
+                self.throw_error(&bytes.to_string());
+                stream.shutdown(Shutdown::Both)?;
+            } else {
+                self.emit(s, &bytes.to_string());
+                stream.shutdown(Shutdown::Both)?;
```

Comment:
The socket its always closed, so no point in doing it inside the conditional blocks.